dnsmasq: fix dnsmasq.init crash-loops with anonymous tags, add match_tag field

systemcrash opened the pull request:
Instead of just exiting and causing dnsmasq to crashloop, prefer the tag name from property "match_tag". Allow anonymous tag sections.

If no match_tag, defer to the tag name, failing that, bail.


@pprindeville Is there some weird shell magic I've overlooked?

pprindeville reviewed (commented):
Some observations...

systemcrash commented on package/network/services/dnsmasq/files/dnsmasq.init:
```diff
@@ -461,16 +461,18 @@ dhcp_tag_add() {
+	config_get match_tag "$cfg" match_tag
+	[ -n "$match_tag" ] || match_tag="$cfg"
+	# if match_tag is empty in anonymous tag entry, bail:
+	[ -z "$match_tag" ] && [ -z "$cfg" ] && return 0
```
Did you mean line 464?

pprindeville replied:
No

systemcrash replied:
Then the first test (`[ -n "$match_tag" ] ||`) on 465 is not redundant. Otherwise how would it help to assign match_tag on 464, then assign it on 465? The assign on 465 is conditional to the first test on 465. 

At least this code is tested and performs as anticipated.

pprindeville replied:
Okay, let's walk through the code.  First, `$cfg` will never be empty, or line 464 won't work.

So by the time the 2nd half of line 465 happens (the assignment), `$match_tag` is not empty.

So line 467 is superfluous because the first and 2nd tests will never be true.

systemcrash replied:
So the assumption is `$cfg` will never be empty. That was the whole problem: it sometimes *is* empty when the tag is anonymous, thus this PR.

pprindeville replied:
Then we shouldn’t call `config_get` if `$cfg` is empty…

systemcrash replied:
Sorry, got my wires crossed. Yeah, $cfg is never empty. 🤦

systemcrash commented:
Just wonder where I'm gonna find another 50 reviews...

systemcrash commented:
Can you review?

@neheb 
@BKPepe

hauke commented:
Which configuration was not working without this change and is working after this change?

systemcrash commented:
Any with anonymous tags. 

Since luci is in common usage, its interface paradigms prefer anonymous sections, so 'named' is minority, unusual or uncommon. 

e.g.
```
config tag
	list dhcp_option '3,192.168.1.201'
	list dhcp_option '28,192.168.1.207'
	list dhcp_option '1,255.255.255.248'
```

systemcrash commented:
Tags need a parameter (here: `match_tag` ) for its coming GUI. luci paradigms do not favour management of names in named entries - where anonymous entries are the main design pattern. Name fields are not manageable by default in luci (requiring delete -> add copy over properties/attributes to new entry).

systemcrash committed:
dnsmasq: dnsmasq.init crash-loops with anonymous tags

Instead of just exiting and causing dnsmasq to crashloop, prefer the
tag name from property "match_tag". Allow anonymous tag sections.

If no match_tag, defer to the tag name, failing that, bail.

Tested on 22.03.5

Signed-off-by: Paul Donald <newtwen+github@gmail.com>
